KGTorrent/mk_preprocessor.py

```diff
@@ -0,0 +1,259 @@
+                    'PrivateScoreFullPrecision'].map(lambda x: round(float(x), 3)).map(
+                    lambda x: x if not np.isinf(x) else np.NaN)
+
+            # DATE COLUMNS FIX
+            date_columns = [column for column in self._tables_dict[table_name].columns if column.endswith('Date')]
+
+            if len(date_columns) != 0:
+                print(f'\t{table_name} parsing date columns...')
+                for column in date_columns:
+                    self._tables_dict[table_name][column] = pd.to_datetime(self._tables_dict[table_name][column],
+                                                                           infer_datetime_format=True,
+                                                                           cache=True)
+
+            # Set initial rows in stats df
+            new_stats_row = {
+                'Table': table_name,
+                'Initial#rows': self._tables_dict[table_name].shape[0],
+                'Final#rows': None,
+                'Ratio': None
+            }
+            self._stats = self._stats.append(new_stats_row, ignore_index=True)
+
+        print()
+
+    def _process_referencing_table(self, referencing):
+        """
+        This method is invoked recursively to drop rows with unsolvable foreign key constraints.
+        It uses :func:`.MKPreprocessor.clean_referencing_table` to do the actual cleaning.
+        Its purpose is to traverse the table relationships graph to take full relationship chains into account.
+        It stops when it detects cycles.
+
+        Args:
+            referencing: the table with foreign keys to be analyzed.
+        """
+
+        print("### PREPROCESSING", referencing)
+
+        self._already_visited.append(referencing)  # TODO: Try to avoid this
+        referenced_list = self._constraints_df.loc[
+            self._constraints_df['Table'] == referencing,
+            'Referenced Table'].values
+
+        # If referenced_list is NOT empty
+        # I recursively process each referenced table and subsequently adjust the current
+        if len(referenced_list) != 0:
+            for referenced in referenced_list:
+                print(self._already_visited)
+                if referenced not in self._already_visited:
+                    self._process_referencing_table(referenced)
+                self._clean_referencing_table(referencing, referenced)
+
+    def _clean_referencing_table(self, referencing, referenced):
+        """
+        Given two tables, a referencing table and a referenced table, this method cleans the referencing table
+        by removing all those rows pointing at a tuple that cannot be found in the referenced table.
+        It also marks the relative foreign key constraint as solved and, whereas the rows are removed from
+        the referencing table, constraints are marked as unsolved where the table referencing is shown as
+        the referenced table in the constraints table
```

Comment:
I cannot understand this docstring.

Reply:
Maybe the word "whereas" is misused here.

Reply:
I use 'relative' instead of 'related' and 'whereas' instead of 'when'. I had difficulties to explain the use of 'IsSolved' field.
Do you have any suggestions to improve the understanding of this part? Or by changing those two words the sentence acquires meaning?

Reply:
This is my suggestion:

> Given two tables - a _referencing table_ and a _referenced table_ - this method cleans the _referencing table_ by removing all the rows that point to missing tuples in the _referenced table_; once the _referencing table_ has been cleaned, this method marks the related foreign key constraint as "solved".
> Moreover, each time the procedure actually removes rows from a _referencing table_, it updates the constraints table accordingly: all the entries in which the current _referencing table_ plays the role of _referenced table_ are marked as "unsolved".